redun/executors/gcp_utils.py

```diff
@@ -224,3 +234,105 @@ def get_task(client: batch_v1.BatchServiceClient, task_name: str) -> batch_v1.Ta
         A Task object representing the specified task.
     """
     return client.get_task(name=task_name)
+
+
+# largely based on nextflow's implementation: https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy
+@dataclass
+class MachineType:
+    type: str
+    family: str
+    spot_price: float
+    on_demand_price: float
+    cpus: int
+    memory: int
+    gpus: int
+
+
+@cache
+def get_available_machine_types(region: str) -> List[MachineType]:
+    CLOUD_INFO_API = "https://cloudinfo.seqera.io/api/v1"
+    request = requests.get(
+        f"{CLOUD_INFO_API}/providers/google/services/compute/regions/{region}/products",
+        timeout=30,
+    )
+    if request.status_code != 200:
+        raise RuntimeError(f"Error getting machine types: {request.text}")
+    result: List[Dict] = request.json()["products"]
+    return [
+        MachineType(
+            type=p["type"],
+            family=p["type"].split("-")[0],
+            spot_price=mean(e["price"] for e in p["spotPrice"]),
+            on_demand_price=p["onDemandPrice"],
+            cpus=p["cpusPerVm"],
+            memory=p["memPerVm"],
+            gpus=p["gpusPerVm"],
+        )
+        for p in result
+    ]
+
+
+def find_best_matching_machine_type(
```

Comment:
shall we include gpus in the request as well or at least note it as a todo for later? it might also be good to filter gpu machine type if no gpus are reuqested

Reply:
Their API doesn't return machines with GPUs for now, added the code for it but would throw an exception if GPUs are requested.